h/streamer/filter.py

```diff
@@ -29,71 +30,98 @@
 
 
 class FilterHandler:
+    FIELD_GETTERS = {
+        "/uri": attrgetter("target_uri"),
+        "/references": attrgetter("references"),
+        "/id": attrgetter("id"),
+    }
+
     def __init__(self, filter_json):
         self.filter = filter_json
 
+        # Pre-compile as much as is sensible about our filters, so we can do
+        # this work once, and then apply it over and over
+        self._clause_filters = [
+            clause_filter
+            for clause_filter in (
+                self._get_filter(clause) for clause in self.filter.get("clauses", [])
+            )
+            if clause_filter
+        ]
+
     def evaluate_clause(self, clause, annotation):
+        clause_filter = self._get_filter(clause)
+
+        if clause_filter is None:
+            return None
+
+        return clause_filter(annotation)
+
+    def include_any(self, annotation):
+        for clause_filter in self._clause_filters:
+            if clause_filter(annotation):
+                return True
+
+        return False
+
+    def match(self, annotation):
+        if not self._clause_filters:
+            return True
+
+        return self.include_any(annotation)
+
+    @classmethod
+    def _get_filter(cls, clause):
+        """Get a filter function for a given clause.
+
+        :param clause: The decoded JSON clause
+        :return: A function which accepts an annotation and returns a truth
+            value depending on whether the value applies
+        """
         field_path = clause["field"]
 
-        # Extract the annotation property that corresponds to the "field"
-        # path in the filter. This only supports a small set of paths
-        # which are actually used by the Hypothesis client.
-        field_value = None
-        if field_path == "/uri":
-            field_value = annotation.target_uri
-        elif field_path == "/references":
-            field_value = annotation.references
-        elif field_path == "/id":
-            field_value = annotation.id
+        try:
+            field_getter = cls.FIELD_GETTERS[field_path]
+        except KeyError:
+            return
 
-        if field_value is None:
-            return False
+        def normalize(term, known_term=False):
+            if not known_term and isinstance(term, list):
+                return [normalize(sub_term, known_term=True) for sub_term in term]
 
-        filter_term = clause["value"]
+            # Notice closure around field_path here
+            if field_path == "/uri":
+                # Apply field-specific normalization.
+                return normalize_uri(term)
```

Comment:
It's hard to see in the diff, but previously we performed `uni_fold` and then discarded the result. This is worth a small but decent uptick in performance by itself.